lti_consumer/lti_consumer.py

```diff
@@ -421,20 +425,83 @@ class LtiConsumerXBlock(StudioEditableXBlockMixin, XBlock):
+               "The processors are site-wide, please consult the site administrator if you have any questions."),
+        default=False,
+        scope=Scope.settings
+    )
 
     # Possible editable fields
     editable_field_names = (
         'display_name', 'description', 'lti_id', 'launch_url', 'custom_parameters',
         'launch_target', 'button_text', 'inline_height', 'modal_height', 'modal_width',
         'has_score', 'weight', 'hide_launch', 'accept_grades_past_due', 'ask_to_send_username',
-        'ask_to_send_email'
+        'ask_to_send_email', 'enable_processors',
     )
 
+    @staticmethod
+    def workbench_scenarios():
+        """
+        Gather scenarios to be displayed in the workbench
+        """
+        scenarios = [
+            ('LTI Consumer XBlock',
+             '''<sequence_demo>
+                    <lti_consumer
+                        display_name="LTI Consumer - New Window"
+                        lti_id="test"
+                        description=""
+                        ask_to_send_username="False"
+                        ask_to_send_email="False"
+                        enable_processors="True"
+                        launch_target="new_window"
+                        launch_url="https://lti.tools/saltire/tp" />
+
+                    <lti_consumer
+                        display_name="LTI Consumer - IFrame"
+                        lti_id="test"
+                        ask_to_send_username="False"
+                        ask_to_send_email="False"
+                        enable_processors="True"
+                        description=""
+                        launch_target="iframe"
+                        launch_url="https://lti.tools/saltire/tp" />
+                </sequence_demo>
+             '''),
+        ]
+        return scenarios
+
     def validate_field_data(self, validation, data):
         if not isinstance(data.custom_parameters, list):
             _ = self.runtime.service(self, "i18n").ugettext
             validation.add(ValidationMessage(ValidationMessage.ERROR, unicode(_("Custom Parameters must be a list"))))
 
+    @property
+    def settings(self):
+        """
+        Gets the XBlock settings bucket via the SettingsService.
+        """
+        settings_service = self.runtime.service(self, 'settings')
+        if settings_service:
+            return settings_service.get_settings_bucket(self)
+
+        return {}
+
+    @property
+    def parameter_processors(self):
+        """
+        Read the parameter processor functions from the settings and return their functions.
+        """
+        if not self.enable_processors:
+            return
+
+        for path in self.settings.get('parameters_processors', []):
+            module_path, func_name = path.split(':', 1)
+            module = import_module(module_path)
+            yield getattr(module, func_name)
```

Comment:
Thanks @ormsbee! Logs makes sense, I'll trap and log errors.

> Any single misconfigured or badly coded parameter processor can't blow up the other properly configured and well behaving processors.

Just to confirm, your suggesting that an error in a processor should be logged as [`log.exception()`](https://stackoverflow.com/a/5191885) and the exception should be suppressed?

This will be tricky to debug for the LTI provider since it will result in parameters being "sometimes magically" vanishing from the LTI parameter list. I find that more obscure to debug, so here's what I suggest as a better solution to address your concerns:

There are two types of errors: configuration errors and processor runtime errors, that would be best to be handled differently:

**Configuration Errors:**
Such as `XBLOCK_SETTINGS` is being set to faulty configs, incorrect module path, providing some weird yaml syntax (e.g. list instead of a dict).

In this case it's not so bad to use `log.exception` for debugging. Then re-raise the exception causing 500 error. Which is much better than failing silently due a DevOps issue, as those error don't just happen, and would probably be best addressed ASAP by platform DevOps.

The error handling would look like the following:
```python
try:
    module_path, func_name = path.split(':', 1)
    module = import_module(module_path)
    yield getattr(module, func_name)
except:
    log.exception('Something went wrong in reading the LTI XBlock configuration for the processor "%s"', path)
    raise
```

**Processor Errors:**
A lot of things can go wrong in a the processor function, i.e. an external service isn't responding, or file access error, etc.

Such errors probably should best not to break the platform. However they should also have a graceful degradation, in addition to logging the error.

So each processor should be able to define a list of expected parameters and their default values. Those parameters are read by the XBlock and be provided as a fallback in case the processor failed. If a processor doesn't specify those parameters, then no default params will be added to the LTI request when an error happens.

A processor with default parameters would look like that:
```python
def user_email_processor(xblock):
    return {
        'email': xblock.user.email,
    }

user_email_processor.lti_xblock_default_params = {
    'email': '',
}
```

Somewhere in the LTI XBlock:
```python
for path, processor in self.xblock.get_parameter_processors():
    default_params = getattr(processor, 'lti_xblock_default_params', {})
    default_params.update(default_params)

    try:
        lti_parameters.update(processor(self.xblock))
    except:
        log.exception('Error in XBlock LTI parameter processor "%s"', path)
```

What do you think?

Reply:
Done.